src/deepsparse/server/build_logger.py

```diff
@@ -0,0 +1,125 @@
+# Copyright (c) 2021 - present / Neuralmagic, Inc. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""
+Specifies the mapping from the ServerConfig to the DeepSparse Logger
+"""
+
+from typing import Any, Dict, List, Union
+
+from deepsparse import loggers as logger_objects
+from deepsparse.loggers.config import MetricFunctionConfig
+from deepsparse.server.config import ServerConfig
+
+
+__all__ = ["build_logger"]
+
+_LOGGER_MAPPING = {"python": logger_objects.PythonLogger}
+
+
+def build_logger(server_config: ServerConfig) -> logger_objects.BaseLogger:
+    """
+    Builds a DeepSparse logger from the ServerConfig.
+
+    The process follows the tree-like hierarchy:
+
+    First: the "leaf" loggers are being built.
+    Second: if specified in ServerConfig (inside the config
+        of every server endpoint; under 'data_logging' value),
+        every "leaf" logger is wrapped inside a FunctionLogger.
+    Third: if required, the list of resulting loggers is wrapped inside a
+        MultiLogger.
+    Fourth: if specified, the resulting logger is wrapped inside the SubprocessLogger
+
+    :param server_config: the Server configuration model
+    :return: a DeepSparse logger instance
+    """
+
+    loggers_config = server_config.loggers
+    if loggers_config is None:
+        return loggers_config
+
+    leaf_loggers = build_leaf_loggers(loggers_config)
+
+    loggers = []
+    for name, leaf_logger in leaf_loggers.items():
+        target_to_metric_function_configs = _get_target_to_metric_function_configs(
+            endpoints=server_config.endpoints, logger_name=name
+        )
+        if target_to_metric_function_configs:
+            logger = logger_objects.FunctionLogger(
+                logger=leaf_logger,
+                target_to_metric_function_configs=target_to_metric_function_configs,
+            )
+        else:
+            logger = leaf_logger
```

Comment:
why are we defaulting to adding a leaf_logger to the top level? this should never be the case - leaf logging should always be under the function logger layer otherwise there will be no control over what is logged (ie all raw data would be passed to the leaf loggers).

let's refactor the function into two parts:
1. build a multi logger of leaf loggers
2. build the dictionary of identifier to metric configs to create a single function logger that wraps the leaf multi logger

Reply:
these are two separate things - do not want to get away from the point that every leaf call in the server environment should exist behind (a/the) function logger

regarding the responsibilities - we need to map and extract values based on targets and run potentially a series of functions on them. we can do this either in an unstructured way with a series of multi loggers and function loggers, or have a single source of truth in a function logger. 

there's merits to both, concern with the former is understanding the composition and flow of the loggers will be a lot tougher and less clear without the structure (imagine being given the top level server logger as a python objet and needing to map out what is going on), while the second could limit flexibility in the future.

Reply:
How about "outsourcing" some responsibilities to MultiLogger? It could potentially be in charge of "turning on/off" the metric functions that override the global loggers. Just a thought. MultiLogger for now does nothing else but serves as a container for multiple loggers (which is its responsibility). But maybe it could do something more. Let's sleep on it.